presto-main/src/main/java/com/facebook/presto/operator/aggregation/AreaUnderROCCurveAggregation.java

```diff
@@ -0,0 +1,217 @@
+            Block labelsBlock = otherState.get().getKeys();
+            Block scoresBlock = otherState.get().getValues();
+            MultiKeyValuePairs pairs = state.get();
+            long startSize = pairs.estimatedInMemorySize();
+            for (int i = 0; i < labelsBlock.getPositionCount(); i++) {
+                pairs.add(labelsBlock, scoresBlock, i, i);
+            }
+            state.addMemoryUsage(pairs.estimatedInMemorySize() - startSize);
+        }
+        else if (state.get() == null) {
+            state.set(otherState.get());
+        }
+    }
+
+    public static void output(MultiKeyValuePairsState state, BlockBuilder out)
+    {
+        MultiKeyValuePairs pairs = state.get();
+        if (pairs == null) {
+            out.appendNull();
+        }
+        else {
+            Block labelsBlock = pairs.getKeys();
+            Block scoresBlock = pairs.getValues();
+            boolean[] labels = new boolean[labelsBlock.getPositionCount()];
+            double[] scores = new double[labelsBlock.getPositionCount()];
+            for (int i = 0; i < labelsBlock.getPositionCount(); i++) {
+                labels[i] = BOOLEAN.getBoolean(labelsBlock, i);
+                scores[i] = DOUBLE.getDouble(scoresBlock, i);
+            }
+            double auc = computeAUC(labels, scores);
+            if (Double.isFinite(auc)) {
+                DOUBLE.writeDouble(out, auc);
+            }
+            else {
+                out.appendNull();
+            }
+        }
+    }
+
+    /**
+     * Calculate Area Under the ROC Curve using trapezoidal rule
+     * See Algorithm 4 in a paper "ROC Graphs: Notes and Practical Considerations
+     * for Data Mining Researchers" for detail:
+     * http://www.hpl.hp.com/techreports/2003/HPL-2003-4.pdf
+     */
+    private static double computeAUC(boolean[] labels, double[] scores)
+    {
+        // labels sorted descending by scores
+        Integer[] sortedIndices = IntStream.range(0, labels.length).boxed().sorted((i, j) -> (int) Math.signum(scores[j] - scores[i])).toArray(Integer[]::new);
+
+        int tp = 0;
+        int fp = 0;
+
+        int tpPrev = 0;
+        int fpPrev = 0;
+
+        double area = 0.0;
+
+        double scorePrev = Double.POSITIVE_INFINITY;
+
+        for (int i : sortedIndices) {
+            if (scores[i] != scorePrev) {
+                area += trapezoidArea(fp, fpPrev, tp, tpPrev);
+
+                scorePrev = scores[i];
+                fpPrev = fp;
+                tpPrev = tp;
+            }
+            if (labels[i]) {
+                tp++;
+            }
+            else {
+                fp++;
+            }
+        }
+
+        // AUC for single class outcome is not defined
+        if (tp == 0 || fp == 0) {
+            return Double.POSITIVE_INFINITY;
+        }
+
+        // finalize by adding a trapezoid area based on the final tp/fp counts
+        area += trapezoidArea(fp, fpPrev, tp, tpPrev);
```

Comment:
the algorithm in the paper does `trapezoidArea(1, fpPrev, 1, tpPrev);` here.

Reply:
I believe we cannot directly use the paper's algorithm. See another paper which describes  the same algorithm: http://people.inf.elte.hu/kiss/13dwhdm/roc.pdf (Algorithm 2); in this second paper, the value is `N` instead of `1` (and I suspect it's also wrong).  

In my understanding, correct algorithm should be: `A <- A + TRAP_AREA(N, FPprev, P, TPprev)`.

This line should add a trapezoidal area between the rightmost "possible" point and the rightmost "observed" point to finalize AUC computation, and `trapezoidArea(1, fpPrev, 1, tpPrev)` means the rightmost possible point is `(1, 1)`. However, since we're simply counting the number of TPs and FPs on the graph, rightmost value on the x-/y-axis should be the total number of negatives and positives in `labels`.

Anyway, I seek more evidence (or counter-example) to show which algorithm is truly correct.